### DISCUSSION TITLE
Multiline prompts

### DISCUSSION THREAD

**Discussion by bcalco** [spectreconsole/spectre.console#683] [Q&A]
I'm using Spectre.Console happily but would like to know how one might go about creating a multiline prompt "widget" that allows for interactive REPL-style coding. The multiple line input would need a termination rule, either some character that escapes the prompt, or otherwise a place to customize that logic so that a variety of different kinds of input termination strategies to be plugged in by the user.

To be more precise, I'm thinking of the interaction experience I get in Chez Scheme, which has a nifty feature of back highlighting left-parens when I type a right parens, even if it was on a previous line. The termination strategy in Scheme (or any Lisp) is balanced parens - so as soon as I hit enter after balancing my final right parens, then the entire multiline response to the prompt is read in and evaluated. Another feature of Chez's REPL editor is the ability to use up/down/left/right to modify the entry like a normal text editor would. I'd like to replicate that kind of experience in a REPL. Bonus points if I can rewrite the input so far and apply syntax highlighting via dynamic highlighting. Another use case is a database REPL where queries are terminated with ';'. 

Any pointers on implementing a widget of this kind (even of lesser ambition) appreciated!

**Comment by phil-scott-78** [comment]
Take a look at the Radline project. I believe this is the direction you are interested in. Any assistance is driving this forward is greatly appreciated!

https://github.com/spectreconsole/radline

**Reply by bcalco** [reply]
Awesome - thanks - I will definitely check that out!

**Reply by bcalco** [reply]
Yes, this is exactly what I'm looking for. How can I assist? I mean aside from using it, which I'm about to do. :)

I see the version at NuGet is 0.5.0 - is that the best way to start using it, or should I check the repo out locally and build the latest version?

**Comment by bcalco** [comment]
OK, I got back to the app in which I wanted to explore RadLine, and saw that it needed upgrading to 0.6.0; however, 0.6.0 only supports .net 6.0, and my app is running (at least right now) as a .NET Framework 4.8 app, so the upgrade failed. That is, it appears that you removed support for .NET Standard 2.0 (which made it usable in .NET Framework). Is this also where Spectre.Console is heading? Or is there some way you can support .NET Framework as well?

**Reply by patriksvensson** [reply]
We will most probably remove `netstandard2.0` support for version 2.0 of Spectre.Console.
Any specific reason you need to stay on .NET Framework for your console applications?

**Reply by bcalco** [reply]
Well, the application my console app compliments runs on Windows-only and the scripting engine I'm calling under the hood has an issue on .NET core since MS removed saving assemblies to file via Reflection.Emit some time ago. So, it only compiles to binary on .NET Framework. Although it certainly can run on .NET core, the issue is libraries take a while to compile in memory on startup, so there is a palpable startup hit. There is also COM interaction on the Windows side, so for all these reasons, I must support .NET Framework in addition to supporting .NET Core.

**Reply by glenn-slayden** [reply]
@bcalco Did you happen to try ditching NuGet and simply building the RadLine library yourself--or even pulling the RadLine .cs sources into your own app--on **.NET Framework 4.8**? If so, please let me know if there were any showstoppers.

And for the RadLine developers, do you happen to know of or recall any serious problems that would prevent making a few necessary tweaks to privately build RadLine on **.NET Framework 4.8**?

possibly related: https://github.com/jonsequitur/dotnet-repl/issues/46

**Comment by crwsolutions** [comment]
For anyone interested I have created `PromptBox` in a demo project (the box under the table):

<img width="635" height="257" alt="image" src="https://github.com/user-attachments/assets/17fd8238-301a-4635-985a-1822652da76d" />

Supports:
- Mulitline (Shift-enter to insert new line)
- Multiline pasting (Ctrl-B, because Ctrl-V insert hard new lines, so input is ended)
- Style customization
- Placeholder
- Indicators at the end of the line (↩) if there is no new line, but it is a space issue.

Sourcecode (You can just copy paste the file in your own project): https://github.com/crwsolutions/NTokenizers.Extensions.Spectre.Console/blob/main/tests/NTokenizers.Extensions.Spectre.Console.ShowCase.Ai/Prompts/PromptBox.cs

Usage:

```csharp
var userPrompt = AnsiConsole.Prompt(new PromptBox() { Placeholder = "Your prompt" });
```

It would be nice to have something like that as part of the library though.
